File: minigit-5/ts/src/commands/symbolic-ref.ts
// symbolic-ref command - Manage symbolic references

import { findRepoRoot } from '../utils';
import { getSymbolicRef, setSymbolicRef, getHead } from '../refs';

export function symbolicRef(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  const positionalArgs: string[] = [];

  for (const arg of args) {
    if (!arg.startsWith('-')) {
      positionalArgs.push(arg);
    }
  }

  if (positionalArgs.length === 0) {
    console.error('usage: minigit symbolic-ref <name> [<ref>]');
    return 1;
  }

  const refName = positionalArgs[0];

  if (positionalArgs.length === 1) {
    // Read mode
    if (refName === 'HEAD') {
      const head = getHead(repoRoot);
      if (head.startsWith('ref:')) {
        console.log(head.slice(5).trim());
        return 0;
      } else {
        console.error('fatal: ref HEAD is not a symbolic ref');
        return 1;
      }
    }

    const target = getSymbolicRef(refName, repoRoot);
    if (target) {
      console.log(target);
      return 0;
    } else {
      console.error(`fatal: ref ${refName} is not a symbolic ref`);
      return 1;
    }
  }

  // Write mode
  const target = positionalArgs[1];

  if (refName === 'HEAD') {
    const { setHead } = require('../refs');
    setHead(`ref: ${target}`, repoRoot);
  } else {
    setSymbolicRef(refName, target, repoRoot);
  }

  return 0;
}

File: minigit-5/py/commands/symbolic_ref.py
"""symbolic-ref command - Manage symbolic references."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from refs import read_head, write_head, get_head_ref


def run(args: list[str]) -> int:
    """Read or update symbolic reference."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    if not args:
        print("usage: minigit symbolic-ref <name> [<ref>]", file=sys.stderr)
        return 1

    ref_name = args[0]

    if len(args) == 1:
        # Read mode
        if ref_name == 'HEAD':
            head_ref = get_head_ref(repo_root)
            if head_ref:
                print(head_ref)
                return 0
            else:
                # HEAD is detached
                print("fatal: ref HEAD is not a symbolic ref", file=sys.stderr)
                return 1
        else:
            print(f"fatal: unknown symbolic ref: {ref_name}", file=sys.stderr)
            return 1
    else:
        # Write mode
        target = args[1]

        if ref_name == 'HEAD':
            write_head(repo_root, f'ref: {target}')
            return 0
        else:
            # For other symbolic refs, we'd need to implement this
            print(f"fatal: symbolic ref update for {ref_name} not supported", file=sys.stderr)
            return 1
